### PR TITLE
Caching pending alarms

### PR DESCRIPTION
**Pull request by peterbryntesson**
feat: Added code and data structures to cache the latest copy of all condition type derived messages for events that have PendingAlarms.IsEnabled flag set.

### CODE REVIEW

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs [hunk 1]
```diff
@@ -735,10 +738,36 @@ internal sealed class SubscriptionWrapper : ISubscription {
                         EndpointUrl = subscription.Session?.Endpoint?.EndpointUrl,
                         SubscriptionId = Id,
                         Timestamp = publishTime,
-                        Notifications = notification.ToMonitoredItemNotifications(
-                                subscription.MonitoredItems)?.ToList()
+                        Notifications = new List<MonitoredItemNotificationModel>()
                     };
 
+                    if (notification != null && notification.Events != null) {
```
Comment (simonjaeger) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs:744]: Could be `notification?.Events != null`

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/PendingAlarmsOptionsModel.cs [hunk 2]
```diff
@@ -49,6 +49,18 @@ public class PendingAlarmsOptionsModel {
                 TimeSpan.FromMilliseconds(SnapshotInterval.Value) : (TimeSpan?)null;
         }
 
+        /// <summary>
+        /// ConditionIdIndex
+        /// </summary>
+        [IgnoreDataMember]
+        public int ConditionIdIndex { get; set; } = -1;
```
Comment (simonjaeger) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/PendingAlarmsOptionsModel.cs:56]: These could be nullable uints. `uint?`. Would be cleaner to check for null than -1.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/PendingAlarmsOptionsModel.cs [hunk 3]
```diff
@@ -49,6 +49,18 @@ public class PendingAlarmsOptionsModel {
                 TimeSpan.FromMilliseconds(SnapshotInterval.Value) : (TimeSpan?)null;
         }
 
+        /// <summary>
+        /// ConditionIdIndex
```
Comment (simonjaeger) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/PendingAlarmsOptionsModel.cs:53]: I know the code looks like this... but can we make these summaries better? A sentence that describes the property.

**Review by simonjaeger — CHANGES_REQUESTED** [review]

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/PendingAlarmsOptionsModel.cs (hunk 2 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa/src/Publisher/Models/Events/PendingAlarmsOptionsModel.cs:56]: Had to make then int?, sinces values[xx] required int. Didn't want to have a cast there

**Review by simonjaeger — APPROVED** [review]

**Comment by simonjaeger** [comment]
LGTM

**Review by mregen — APPROVED** [review]

**Pull request merged**